pkg/inspector/image-inspector.go

```diff
@@ -177,25 +179,94 @@ func (i *defaultImageInspector) Inspect() error {
 }
 
+// aggregateBytesAndReport sums the numbers recieved from its input channel
+// bytesChan and prints them to the log every PULL_LOG_INTERVAL_SEC seconds.
+// It will exit after bytesChan is closed.
+func aggregateBytesAndReport(bytesChan chan int) {
+	var bytesDownloaded int = 0
+	ticker := time.NewTicker(PULL_LOG_INTERVAL_SEC * time.Second)
+	defer ticker.Stop()
+	for {
+		select {
+		case bytes, open := <-bytesChan:
+			if !open {
+				return
+			}
+			bytesDownloaded += bytes
+		case <-ticker.C:
+			log.Printf("Downloading Image (%dKb downloaded)", bytesDownloaded/1024)
+		}
+	}
+}
+
+// decodeDockerPullMessages will parse the docker pull messages received
+// from reader and will push the difference of bytes downloaded to
+// bytesChan. After reader is closed it will close bytesChan and exit.
+func decodeDockerPullMessages(bytesChan chan int, reader io.Reader) {
+	type progressDetailType struct {
+		Current, Total int
+	}
+	type pullMessage struct {
+		Status, Id     string
+		ProgressDetail progressDetailType
+	}
+	defer func() { close(bytesChan) }()           // Closing the channel to end the other routine
+	layersByetsDownloaded := make(map[string]int) // bytes downloaded per layer
+	dec := json.NewDecoder(reader)                // decoder for the json messages
+	for {
+		var v pullMessage
+		if err := dec.Decode(&v); err != nil {
+			if err != io.ErrClosedPipe {
+				log.Printf("Error decoding json: %v", err)
+			}
+			break
+		}
+		// decoding
+		if v.Status == "Downloading" {
+			bytes := v.ProgressDetail.Current
+			last, existed := layersByetsDownloaded[v.Id]
+			if !existed {
+				last = 0
+			}
+			layersByetsDownloaded[v.Id] = bytes
+			bytesChan <- (bytes - last)
+		}
+	}
+}
+
 // pullImage pulls the inspected image using the given client.
 // It will try to use all the given authentication methods and will fail
 // only if all of them failed.
 func (i *defaultImageInspector) pullImage(client *docker.Client) error {
 	log.Printf("Pulling image %s", i.opts.Image)
-	imagePullOption := docker.PullImageOptions{Repository: i.opts.Image}
 
 	var imagePullAuths *docker.AuthConfigurations
 	var authCfgErr error
 	if imagePullAuths, authCfgErr = i.getAuthConfigs(); authCfgErr != nil {
 		return authCfgErr
 	}
 
+	reader, writer := io.Pipe()
+	// handle closing the reader/writer in the method that creates them
+	defer writer.Close()
+	defer reader.Close()
+	imagePullOption := docker.PullImageOptions{
+		Repository:    i.opts.Image,
+		OutputStream:  writer,
+		RawJSONStream: true,
+	}
+
+	bytesChan := make(chan int)
+	go aggregateBytesAndReport(bytesChan)
+	go decodeDockerPullMessages(bytesChan, reader)
+
 	// Try all the possible auth's from the config file
 	var authErr error
-	for _, auth := range imagePullAuths.Configs {
+	for name, auth := range imagePullAuths.Configs {
 		if authErr = client.PullImage(imagePullOption, auth); authErr == nil {
 			return nil
 		}
+		log.Printf("Authentication with %s failed: %v", name, authErr)
```

Comment:
Yes, thanks.   That will at least log something for us if the PullImage has an error that is not auth related.